# Python Basics, Part 1

In our first session this morning, we want to get you familiar with how to use `JupyterLab`, along with the basics of `Python` syntax and the data types we use in `Python`: numbers, strings, lists, and dictionaries. If you have used `Python` before, some of this material will be review, but you'll probably see some new stuff too!

## -1. What is Python?

`Python` is a flexible and widely used programming language in data science and software engineering. Unlike many programming languages, one of the central tenets of the design (and writing) of `Python` is that `Python` code should be _easy to read_. Python is frequently used in a wide range of applications, from web development to machine learning.

Why do people like to use `Python`? `Python` is:
* Fast,
* Modular,
* Object-oriented,
* Extensible (i.e., there are _lots_ of libraries),
* Easy to read,
* Most importantly: gets the job done.

`Python` is the name of a programming language. But different people often interface with `Python` in a number of different ways:
1. Running interactive commands with the `python` interpreter. This is done by typing in `python` (or `python3`, if you have `Python 2` and `Python 3` installed on your computer) at the command line.
<img src=img/ss_python_interp.png width=500>
2. Python development in some kind of text editor or integrated development environment (IDE).
<img src=img/ss_python_spyder.png width=500>
3. Research-related scripting with heavy documentation and snippets of code. This is especially common in data wrangling.
<img src=img/ss_python_notebook.png width=500>
    
Today, we'll be focusing on the third approach, namely, using `Python` in `Jupyter` notebooks for research, but it's worth exploring all approaches!

## 0. Using `Jupyter` Notebooks

`Jupyter` notebooks are an excellent tool for integrating your code, notes and explanations, and results all in one place. `Jupyter` notebooks are reproducible, and allow you to capture not only _what_ you did in your analysis, but also _why_ you did the analysis. We'll be working with them through the `JupyterLab` interface.

### 0.0. Launching `JupyterLab`

Getting a `Jupyter` notebook up and running is easy. Simply open up your favorite terminal app, and type `jupyter lab`. `Jupyter` will print some information about the server (which you can ignore for now), and then a webpage should open in your browser that looks like the second picture below.

<img src="img/ss_jupyter_launch.png" width=500>

<img src="img/ss_jupyter_main_screen.png" width=500>

(Don't worry if things look slightly different locally---the appearance of your terminal and the `JupyterLab` interface will change slightly from computer to computer.)

#### 0.0.0.Exercise

Open `JupyterLab` on your computer.

### 0.1. Navigation

Navigating around `JupyterLab` is relatively easy. The screen is broken into three pieces:

1. **The Menu Bar:** Similar to a program like Microsoft Word, many important actions are performed through the menu bar located at the top of the screen. To modify, save, or close files, or navigate through directories, click, e.g., `File > Save`. To alter the appearance of the `JupyterLab` interface, click, e.g., `View > Show line numbers`.
2. **The Side Bar:** The left sidebar is home to a number of commonly used tabs, the names of which you can display by mousing over them. We will primarily work with the topmost tab, which contains the file browser. Other tabs allow you to manage kernel and terminal sessions, issue commands, or control open tabs.
3. **The Main Panel:** This is the main surface for programming and other activities. The main panel is broken up into tabs, which can be dragged, dropped, and resized.

<img src="img/ss_screen_example.png" width=500>

In the picture above, the main panel is overlaid in blue, the sidebar is overlaid in red, and the menu bar is overlaid in green.

To launch a notebook, simply navigate to the folder containing the notebook, and double-click it.

#### 0.1.0. Exercise

Open `python_basics_part_1.ipynb` to start the first lesson. You should now be able to see locally the instructions that I've been projecting onto the main screen.

### 0.2. Markdown and Code Cells

As we mentioned earlier, `Jupyter` notebooks are a valuable tool because they allow you to interweave explanatory prose for humans to read with code for your computer to run. The way this is accomplished is with "cells." Cells come in two types: "text" or "markdown" cells, where your notes and explanations live; and "code" cells, which will contain your `Python` code.

Cells can be created by clicking on the "+" icon located at the top of the current pane.

<img src=img/ss_add_cell.png width=500>

#### 0.2.0. Exercise

Try adding a cell to this notebook with the "+" icon. To remove it, click on the scissors icon next to it. (This actually "cuts" the cell, so it will still be present on your clipboard.)

#### 0.2.1. Markdown Cells

Text cells are formatted using a lightweight markup language called `markdown`. What this means is that you can easily style the text.

* For *italics*, write `*text*` or `_text_`.
* For **bold** text, write `**text**` or `__text__`.
* To add bulleted items, simply add an asterisk to the beginning of a line, e.g., `* text text text`.
* To start a numbered list, begin the line with a number and a period, e.g., `1. text text text`.
* To insert `code` directly into a text block, simply place backticks around the code, e.g., `` `code` ``.

While these basics are enough for our purposes, much, much more is possible in `markdown`: you can add headers, footnotes, and even directly insert HTML. For a more thorough introduction, click [here](https://www.markdownguide.org/getting-started).

#### 0.2.2. Code Cells

The bread and butter of `Jupyter` notebooks are code cells, which we'll be working with at length below. Code cells have two parts: an input box, where you write the `Python` code, and an output box.

<img src="img/ss_code_cell.png" width=500>

Here, the input is higlighted in red, and the output is highlighted in blue.

#### 0.2.3. Exercise

Create a cell, and then change it from a code cell to a markdown cell and back again. Use the dropdown menu located  in the toolbar:

<img src="img/ss_change_cell.png" width=500>

### 0.3. Saving and switching between notebooks

Let's say you've added a lot of code and text to a notebook---what should you do to "render," or complete the notebook? Click `Run > Run all cells`. This will output nicely formatted text in all of the text boxes, and run all of the code in the code cells. You can also click the play button in the current pane.

To save the notebook output, go to `File > Save`.

### 0.4. How to get help

Even seasoned software engineers frequently come across functions they don't know how to use or code they don't understand. One of the best places to look for help is at the official `Python` documentation. To bring it up, simply click `Help > Python Reference` in the menu bar. Usually, you can find the function (or "module," `Python`'s special word for libraries) just by searching in the search box!

#### 0.4.1. Exercise

Python also has a built-in help function: `help()`. It can be a little intimidating at first, but it's one of the best ways to get help on whatever you're working with. Practice calling the `help()` function by calling `help()` on, well, itself: `help(help)`.

In [101]:
### START
help(help)
### END

Help on _Helper in module _sitebuiltins object:

class _Helper(builtins.object)
 |  Define the builtin 'help'.
 |  
 |  This is a wrapper around pydoc.help that provides a helpful message
 |  when 'help' is typed at the Python interactive prompt.
 |  
 |  Calling help() at the Python prompt starts an interactive help session.
 |  Calling help(thing) prints help for the python object 'thing'.
 |  
 |  Methods defined here:
 |  
 |  __call__(self, *args, **kwds)
 |      Call self as a function.
 |  
 |  __repr__(self)
 |      Return repr(self).
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)



## 1. Warmup: Approximating Pi

Let's begin with a warmup: using `Python` to approximate $\pi \approx 3.14159\ldots$.

There's a smart aleck way to do this: `Python` already stores the constant `pi` in the `math` module:

In [1]:
import math

math.pi

3.141592653589793

(Don't worry if none of that makes sense yet. It all will by the end of the day today!)

Let's instead pretend that we need a way of calculating $\pi$ on our own. From scratch.

How might we go about doing that? Well, one thing that we definitely know about $\pi$ is the classic formula for the area of a circle: $A_\circ = \pi \cdot r^2$. The area of a square is even easier: $A_\square = h^2$. Let's look at a picture:

<img src=img/ss_circ_sqr.png width=500>

Clearly the blue circle has area $\pi\ \text{units}^2$ and the red square has area $4\ \text{units}^2$. What's behind the shapes?

<img src=img/ss_dart_board.png width=500>

A dart board! What's interesting about this dartboard is the probabilities of hitting it. Let's imagine we're playing darts and our aim is ok, but not very good. We can always land in the red square, but within the red square we're as likely to hit one spot as another.

#### 1.0. Exercise
What percentage of our darts should we expect even to hit the board?

In [2]:
# Type your answer here!

So if we just had a dartboard instead of this pesky `JupyterLab` notebook, approximating $\pi$ would be a breeze! We'd just throw lots of darts at the board. We'd just count the percentage that hit the board, instead of missing, and then multiply by four.

Fortunately, this strategy---frequently called _Monte Carlo simulation_---works in `Python` too!

Let's begin by generating some dart throws. A dart throw needs an $x$ and a $y$ coordinate. Our darts always land in the box, so we need the darts coordinates to be between $1$ and $-1$; but other than that, they should be completely random.

Python has a whole bunch of functions for doing random things all ready for us. One of these, `uniform(a,b)`, will give us a randomly generated number between `a` and `b`.

#### 1.1. Exercise
How could we use this to generate the $x$ coordinate of a dart throw? Try it out in the box below:

In [3]:
from random import uniform

# START
uniform(-1,1)
# END

-0.906214449967071

Cool! But that's only one coordinate---what about the $y$ coordinate? Remeber, we need the point $(x,y)$ where the dart lands. This notation is pretty suggestive of how things are done in `Python`: if we want to store a pair of numbers, just write `(NUMBER_1, NUMBER_2)`.

#### 1.2. Exercise
In the code box below, try and generate _both_ coordinates of a random dart throw, and put them in a pair.

In [4]:
# START
(uniform(-1,1), uniform(-1,1))
# END

(0.9523076808185658, -0.3576325930548374)

That's only one throw, though. To really approximate $\pi$, we need _lots_ of throws. Fortunately, `Python` makes it very easy to do something repeatedly: that's what the `for` keyword is for.

Let's suppose we want to model a busy graduate student's schedule across ten days. We could write what's called a _`for` loop_ to do it:

In [5]:
for day in range(0,10):
    print("On day", day, "I worked.")

On day 0 I worked.
On day 1 I worked.
On day 2 I worked.
On day 3 I worked.
On day 4 I worked.
On day 5 I worked.
On day 6 I worked.
On day 7 I worked.
On day 8 I worked.
On day 9 I worked.


Here `Python` simulated (very loosely) what the graduate student did on each of the ten days. If we wanted to simulate fifteen days instead, or only want to know what the graduate student did on days five through eight (inclusive), we could do that too by changing the `range` of our `for` loop:

In [6]:
for day in range(0,15):
    print("On day", day, "I worked.")

print()
    
for day in range(5,9):
    print("On day", day, "I worked.")

On day 0 I worked.
On day 1 I worked.
On day 2 I worked.
On day 3 I worked.
On day 4 I worked.
On day 5 I worked.
On day 6 I worked.
On day 7 I worked.
On day 8 I worked.
On day 9 I worked.
On day 10 I worked.
On day 11 I worked.
On day 12 I worked.
On day 13 I worked.
On day 14 I worked.

On day 5 I worked.
On day 6 I worked.
On day 7 I worked.
On day 8 I worked.


__PROTIP:__ You might notice that our for loops (and even our section numbering) begin with 0 instead of 1. This is because `Python` is what's called a _zero-indexed_ programming language. In general, that means that the _first_ thing in any sort of `Python` object isn't actually 1---it's 0!

Let's say that the graduate student's advisor wants to know what they were up to for the last few days. The graduate student needs to _record_ something showing what they were doing each day. Fortunately, there's an easy pattern to do this too:

In [7]:
record = []
for day in range(0,10):
    record.append("On day " + str(day) + " I worked.")

record

['On day 0 I worked.',
 'On day 1 I worked.',
 'On day 2 I worked.',
 'On day 3 I worked.',
 'On day 4 I worked.',
 'On day 5 I worked.',
 'On day 6 I worked.',
 'On day 7 I worked.',
 'On day 8 I worked.',
 'On day 9 I worked.']

#### 1.3. Exercise
Use the pattern above to store 10 random dart throws.

__PROTIP:__ You should rewrite the code yourself instead of copying and pasting! In general, copying and pasting code is something you want to avoid: when you write code out yourself, you get more familiar with how it works, and copying and pasting will often introduce errors that are annoying and time-consuming to fix. Later on we'll talk about _functions_ and other good ways to avoid repeating yourself, but for now just keep in mind that you should be writing things out yourself.

__HINT:__ If you try to do something like `record.append(a,b)` you might get an error saying `append()` takes exactly one argument. `Python` thinks you're trying to tell it to `append()` two things, and it only knows how to append one. If you combine those two things together into one thing first, it'll work correctly: `append((a,b))`.

In [8]:
# START
record = []
for day in range(0,10):
    record.append((uniform(-1,1), uniform(-1,1)))

record
# END

[(0.12139507306876673, -0.32262559721332873),
 (-0.005635493314294804, 0.32649961819488804),
 (0.3540985836286832, -0.8929088695376965),
 (-0.2098344904342453, -0.553477455433004),
 (-0.1967434114791995, -0.23994255673885),
 (-0.5005175563302904, 0.596487667571191),
 (0.4407251697059984, 0.7628578530326322),
 (-0.32566688043065706, 0.5748241054738785),
 (-0.6366935506432818, 0.6603360739889139),
 (0.3576114388970828, -0.27116190352945124)]

We're very nearly there! The only thing we still have to do is figure out if a point is in the circle or not. We've written a function for you called `on_board()`, which will check to see if a dart throw lands on the board or not. Don't worry if it looks a little complicated---there's nothing going on that you won't understand by the end of the day! 

In [9]:
from math import sqrt

def on_board(x, y):
    if (abs(x) > 1 or abs(y) > 1):
        raise Exception(
            'Your throw should lie in the square from (-1, -1) to (1, 1). The'
            ' throw you gave me was: ' + str((x, y))
        )
    if (sqrt(x ** 2 + y ** 2) < 1):
        return True
    else:
        return False

#### 1.4. Exercise

Using `on_board()` and the `for` loop pattern we just learned, simulate 1,000,000 test throws at the dart board, but instead of recording _where_ the dart landed (i.e., it's coordinates), record _whether_ the dart landed on the board.

In [102]:
### START
record = []
for throw in range(0, 1000000):
    record.append((on_board(uniform(-1, 1), uniform(-1,1))))
### END

Now, all that's left to calculate $\pi$ is to find the percentage of times our dart hit the board and multiply by four!

#### 1.5. Exercise
Use your 1,000,000 simulated dart throws to approximate $\pi$.

__HINT:__ If you need to calculate the percentage of times something happens, use the `mean()` function. E.g.,
```python
    >>> mean([1,2,3])
    2
    >>> mean([True, True, True])
    1
    >>> mean([True, False, False])
    0.3333333333333333
```
note that _before you can use `mean()`_ you have to import it from the `statistics` module. That means you need to write the following line at the top of your code cell:
```python
    from statistics import mean
```

In [103]:
from statistics import mean

### START
mean(record) * 4
### END

3.139936

Amazing! Admittedly, you probably know a better approximation of $\pi$ if you just remember that International Pi Day is March 4th, but the technique we just used---Monte Carlo simulation---is an incredibly powerful technique, especially in situations where we don't know the answer ahead of time.

__PROTIP:__ We actually could increase the accuracy of our simulation just by throwing more darts. The difficulty is that the accuracy of our simulation looks like $1/\sqrt{n}$ (do you see why?), so to get an extra digit of accuracy, we have to increase the number of throws by a factor of 100. You can put `%%timeit` at the beginning of a code cell to see how long it takes to run. Go back to Exercise 0.4 and see how long it takes to simulate 1,000,000 throws. About how many throws would we need for three digits of accuracy? Four?

## 2. Numbers

That was an awful lot of stuff we just did. Let's try and unpack everything that happened, starting with how we did (a little bit of) stuff with numbers.

### 2.0. Arithmetic

The answer is that you probably _already_ know how to do lots of things in `Python`! One useful feature of `Python` that I use quite often is that it can be used as a calculator. Let's try typing in a basic numerical formula.

In [12]:
1 + 1 # Try something like: `1 + 1`

2

Cool! Probably not something we really needed an expensive computer to figure out, but good to know in any case. You can make `Python` do all sorts of numerical operations for you.

#### 2.0.0. Exercise

Get the hang of using `Python` as a calculator. See if you can figure out how to do the following: addition, subtraction, multiplication, division, exponentiation, _integer_ division, (i.e., $5 / 3 = \mathbf{1}\ \text{remainder}\ 2$), and _remainder_ (i.e., $5 / 3 = 1\ \text{remainder}\ \mathbf{2}$).

In [106]:
### START
1 + 1
4.2 - 1.8
2 * 3
4 / 9    
3 ** 2
5 // 2
5 % 2
### END

1

In addition to the standard `+`, `-`, `*`, and `/`, there are a couple operations here that may seem unfamiliar:
* `**` is the exponentiation operator, so `2 ** 3` returns `8`.
* `//` is the _integral_ division operator, so `5 // 2` returns `2` (with a remainder of `1`).
* `%` is the _modular_ ("mod") division---or _remainder_---operator, so `5 % 2` returns `1`. Modular division is very useful when programming. For instance, if you want to do something "every other time," you can simply do it every time `i % 2 == 0`. (More on this later.)

### 2.1. Comparison operators

It's also possible to make `Python` tell you whether certain things are true or false, using the _less than_ (`<`), _greater than_ (`>`), and _equals_ (`==`) operators.

In [20]:
1 > 2

False

In [21]:
2 > 1

True

In [22]:
(2 ** 3) ** 3 == 2 ** 9

True

### 2.1.0. Exercise

Try a couple of comparisons of your own. Is $e^{\pi}$ less than, greater than, or equal to $\pi^e$?

In [107]:
### START
2.71828 ** 3.14159 > 3.14159 ** 2.71828
### END

True

**A WORD OF CAUTION:** Note that the equality operator is `==`---i.e., _two_ equals signs---and not `=`---i.e., a single equals sign. The single equals sign, `=`, is for assigning to variables, as we saw in the warmup, and which we'll talk about in a second. This is a very commn source of errors; even expert programmers will inadvertently type `=` when they meant to type `==`.

__PROTIP:__ In addition, you might have noticed that numbers in `Python` come in two types: _integers_, which are just whole numbers; and _floating point numbers_, which can represent decimal numbers, like `1.5`, `3.14159`, or `2019.0`. (You'll notice that we're representing a whole number using floating point in `Python`, the first decimal place is still printed.)

The important difference between integers and floating point numbers is that operations on integers are _exact_, whereas operations on floating point numbers are only _approximate_. For instance, compare the following:

In [23]:
(((((2 ** 10) // 2 ** 5) * 2 ** 5) // 2 ** 5) * 2 ** 5) == 2 ** 10

True

In [24]:
1.2 - 1.0 == 0.2

False

## TODO(jgaeb): Put control flow here

## 3. Variables

Let's suppose we're running a grocery store, and we have some inventory:
* __Milk,__ which we bought for \\$0.99 per gallon, and resell for \\$1.99 per gallon.
* __Eggs,__ which we bought for \\$2.59 per dozen, and resell for \\$4.00 per dozen.
* __Bread,__ which we bought for \\$6.31 a loaf, and resell at \\$7.99.
* __Coffee,__ which we bought for \\$0.31 a cup, and resell for \\$2.99 a cup.

Alright, now let's suppose it's been a slow morning: we sold four gallons of milk, 36 eggs, two loaves of bread, and seven cups of coffee. What was our net profit?

This probably seems like an annoying problem, and it would be very annoying to solve given what we know how to do in Python so far. We could try typing something like

In [25]:
(1.99 * 4 + 2.59 * 3 + 7.99 * 2 + 2.99 * 7) - (0.99 * 4 + 4.00 * 3 + 6.31 * 2 + 0.31 * 7)

21.89

Not only was this gross to do, but it's not even right. I accidentally introduced an error into the calculation, and didn't get the profit right.

In long calculations, it can become difficult to keep track of intermediate steps. Luckily, `Python` allows us to store the intermediate results of calculations.

### Storing values in variables

A better way to solve this problem is to store intermediate results in variables. This is done using `=`, i.e., the _single_ equals sign. For instance, we might do something like

In [26]:
price_milk_buy    = 0.99
price_milk_sell   = 1.99
price_eggs_buy    = 2.59
price_eggs_sell   = 4.00
price_bread_buy   = 6.31
price_bread_sell  = 7.99
price_coffee_buy  = 0.31
price_coffee_sell = 2.99

revenue = (price_milk_sell * 4 +
           price_eggs_sell * 3 +
           price_bread_sell * 2 +
           price_coffee_sell * 7)

costs   = (price_milk_buy * 4 +
           price_eggs_buy * 3 +
           price_bread_buy * 2 +
           price_coffee_buy * 7)

revenue - costs

30.349999999999994

### Displaying variables with `print()`

Let's suppose that we wanted to save our final computation---that is, our net profit---for later use. We might do something like

In [27]:
profit = revenue - costs

This is nice, but you'll notice that the previous code cell doesn't have any output. How do we figure out what our profit actually _was_? This is where the `print()` function comes in handy.

In [28]:
print(profit)

30.349999999999994


The nice thing about `print()` is that it gives us a lot of control over how numbers are displayed. For instance, if all we saw from the last cell were the output---i.e., `30.349999999999994`---it would be pretty hard to figure out what was going on. Luckily, it's fairly easy to augment this bare number:

In [29]:
print("Our profit this morning was", profit)

Our profit this morning was 30.349999999999994


We can do this with many arguments:

In [30]:
name = "Hans"
time = "morning"
print("Hello, my name is", name, "and I'll be going over some Python basics this", time)

Hello, my name is Hans and I'll be going over some Python basics this morning


### Things to watch out for

Here are two things to remember to avoid some variable-related headaches down the road.

#### `Python` variables are case sensitive

In `Python`, variables are case sensitive. That means that `dog` and `DOG` possibly represent different things.

In [31]:
dog = "good boy"
DOG = "WOOF"
dog == DOG

False

#### Variables have to be stored before they can be used

If we try to bring our neighbor's dog into the equation without first saying what it is, `Python` will throw an error.

In [32]:
dog == DoG

NameError: name 'DoG' is not defined

Some languages will try to infer a "default value," but `Python` will not. However, there's no issue if we store a value first and then use it.

In [33]:
Dog = "neighbor dog"
dog == Dog

False

## Strings

You might have noticed that some of the variables we used in the last actually stored text instead of numbers. Text, or "strings," as they're usually called, are one of the things that can be a bit of a hassle in other languages, but are very easy to deal with in `Python`.

Strings are enclosed in either single quotes(`'...'`) or double quotes (`"..."`). The only difference betwen the two is that you need to escape literal double quotes with `\` if you're using single quotes, and vice versa. Examples:

In [34]:
'This is a string in single quotes, and it works fine'

'This is a string in single quotes, and it works fine'

In [35]:
'This is a string in single quotes, but it's broken!'

SyntaxError: invalid syntax (<ipython-input-35-ff88f75c76eb>, line 1)

In [36]:
'This is a string with single quotes, and now it\'s fixed!'

"This is a string with single quotes, and now it's fixed!"

In [37]:
"So, there's incentive to use double quotes"

"So, there's incentive to use double quotes"

That should make it pretty clear. Note that when escaping a single quote in a string enclosed in single quotes, the interpreter internally changes the enclosing quotes to double quotes. The two are absolutely identical, so you should use whichever set of quotes you prefer to enclose your strings, as long as you're consistent.

Use the `print` function to make output more readable by omitting the enclosing quotes and printing special characters escaped with `\`.

In [38]:
'enclosing double quotes ("") and single quotes(\'\') are the same thing in python'

'enclosing double quotes ("") and single quotes(\'\') are the same thing in python'

In [39]:
print('enclosing double quotes ("") and single quotes(\'\') are the same thing in python')

enclosing double quotes ("") and single quotes('') are the same thing in python


Stings can also have more than one line. You can either use an explicit line-break character 
(`\n`):

In [40]:
print('This string has\ntwo lines!')

This string has
two lines!


... or use triple quotes `'''...'''` or `"""..."""`

In [41]:
print('''
This string has
two lines!
''')


This string has
two lines!



If you look carefully enough, you'll notice that the last string actually has four lines. This is because the triple quotes literally encode all white spaces, including the new lines after the first `'''` and the last "`!`". To avoid this, you can escape the new lines with `\`.

In [42]:
print('''\
This string (really) has
two lines!\
''')

This string (really) has
two lines!


What if you want to write a string that actually contains the `\` character?
You can either:
* escape `\` with `\` (e.g., write **two** `\` characters for one), or
* prepend a single `r` to the quotes to indicate that you are writing a **r**aw string

In [43]:
print("A backslash (\\) is awesome!")
print(r"A backslash (\) is awesome!")

A backslash (\) is awesome!
A backslash (\) is awesome!


Unlike some other stingy languages, the plus operator (`+`) does exactly what you'd expect it to do with strings!

In [44]:
first_name = 'Hans'
last_name = 'Gaebler'
full_name = first_name + ' ' + last_name
print('Hello', full_name)

Hello Hans Gaebler


Even the multiplying operator (`*`) works!

In [45]:
print('Sing, ' + 'la ' * 3)

Sing, la la la 


## Lists

Strings are great and all, but they're hardly the most versatile data type in `Python`. In particular, while they're helpful for storing _textual_ data, they aren't very useful for dealing with much else. Luckily, `Python` has lists!

Lists are "array-like." What that means is that lists consist of lots of slots, each slot containing some item. The easiest way to construct lists is by placing the items you want in between square brackets like so: `[item_1, item_2, item_3]`. For instance,

In [46]:
fruits = ["apples", "bananas", "pears", "mangosteens", "strawberries"]

print(fruits)

['apples', 'bananas', 'pears', 'mangosteens', 'strawberries']


Lists can contain more than just strings:

In [47]:
# A list containing numbers
primes = [2, 3, 5, 7, 11, 13]

# A list containing other lists
junk_drawer = [fruits, primes, ["kittens", "kettles", "mittens", "packages"]]

# A list of different things
misc = [3, 4, "five", "six"]

print(primes, junk_drawer, misc)

[2, 3, 5, 7, 11, 13] [['apples', 'bananas', 'pears', 'mangosteens', 'strawberries'], [2, 3, 5, 7, 11, 13], ['kittens', 'kettles', 'mittens', 'packages']] [3, 4, 'five', 'six']


That's all well and good if we want to put things _into_ lists, but how do we take things _out of_ lists? The answer is that lists (and in fact, all built-in `python` [sequence](https://docs.python.org/2/glossary.html#term-sequence) types) can be indexed and sliced.
* __Indexing__ refers to when we get a specific element of a list at a given position. The result is an _element_ of a list, which may or may not be a list itself.

In [48]:
print(fruits[0])      # Returns the fruit at position 0
print(junk_drawer[1]) # Returns the _list_ at position 1

apples
[2, 3, 5, 7, 11, 13]


* __Slicing__ refers to when we take a "slice" of a list between some given indices. The result is _always_ another list.

In [49]:
misc[1:3]

[4, 'five']

Also like strings, lists can be concatenated with the `+` operator:

In [50]:
fruits + ["tomatoes", "grapefruits"]

['apples',
 'bananas',
 'pears',
 'mangosteens',
 'strawberries',
 'tomatoes',
 'grapefruits']

## Strings are (almost) like lists

Strings, like many things in `python`, can be indexed (subscripted). The first element (character) has index 0.

In [51]:
job = 'jedi'
job[0]  # character at position 0

'j'

`Python` will yell at you if you go out of range, whether you do so with a string or a list

In [52]:
job[10]

IndexError: string index out of range

In [53]:
fruits[7]

IndexError: list index out of range

But you *can* go backwards with a negative index, -1 being the right-most character.

In [54]:
print(job[-1])    # right-most character
print(fruits[-1])  # right-most fruit

i
strawberries


Like with lists, *slicing* is another useful way to get subsets of your string. 

In [55]:
job[0:2]  # characters from position 0 (included) to 2 (excluded)

'je'

In [56]:
job[2:4]  # characters from position 2 (included) to 3 (excluded)

'di'

Omitting the first slice index will default to zero (the first element)

In [57]:
print(job[1:])    # slice from position 1 (included) to the end
print(fruits[1:])

edi
['bananas', 'pears', 'mangosteens', 'strawberries']


Use slices creatively, to make your life easier!

In [58]:
job[-2:]  # slice the last two characters

'di'

Unlike indexing, slicing is generous to ambitious ranges

In [59]:
job[0:100]

'jedi'

The big difference in `Python` between strings and lists is what's called "mutability." Strings in `Python` are _immutable,_ meaning they can't be changed. In other words, you can't assign a value to a string index.

In [60]:
job[0] = 'T'

TypeError: 'str' object does not support item assignment

Instead, you have to build a new string from the existing string.

In [61]:
job = 'T' + job[1:]
print(job)

Tedi


Lists, however, can be modified at a certain value. This makes them _mutable_. For instance, 

In [62]:
misc[0] = 30  # change indexed item
print(misc)

[30, 4, 'five', 'six']


### Convenience functions for lists and strings

Strings and lists have a lot in common. In particular, there are certain things that we do with both of them _so often_ that `Python` provides a whole bunch of functions to do the job for us. For instance, it's common to want to know the length of both lists and strings. That's where the `len()` function comes in:

In [63]:
print("The length of `fruits` is", len(fruits))
print("The length of `job` is", len(job))

The length of `fruits` is 5
The length of `job` is 4


There are a _whole bunch_ of these functions. We're only going to go through a few of them here, but you can see all of them in the docs [here](https://docs.python.org/3/library/stdtypes.html#string-methods) for strings and [here](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists) for lists.

Don't worry if these don't all sink in right away. The point is just to give you a taste of what's possible. As a general rule, if it's easy to describe what you want to do, and you could do it to any list (or string), there's probably already a built-in function for doing it.

In [64]:
s = 'Hi, my name is Johann. Yes, Johann.'
s.split()  # splits the string into a list (split at spaces by default)

['Hi,', 'my', 'name', 'is', 'Johann.', 'Yes,', 'Johann.']

In [65]:
s.split(',')  # can specify which character to split at

['Hi', ' my name is Johann. Yes', ' Johann.']

In [66]:
s.count('n')  # count the number of non-overlapping occurrences of a substring

5

In [67]:
s.count('Jongbin')

0

In [68]:
s.upper()  # makes everything uppercase

'HI, MY NAME IS JOHANN. YES, JOHANN.'

In [69]:
s.lower()  # makes everything lowercase

'hi, my name is johann. yes, johann.'

In [70]:
s.lower().count('y')  # methods that return a string can be chained

2

In [71]:
':'.join(s.split())

'Hi,:my:name:is:Johann.:Yes,:Johann.'

That last one is a little tricky. So, `str.join(some_sequence)` will take each item of `some_sequence` and stick them together with the value of `str` inbetween, making a single large string. It may seem like a crazy thing to do, but is actually pretty useful whem converting data into comma-separated values. i.e.,

In [72]:
','.join(['some','data','in','a','list'])

'some,data,in,a,list'

You can also use the <kbd>Tab</kbd> character (`\t`) to create tab-separated values:

In [73]:
print('\t'.join(['some','data','in','a','list']))

some	data	in	a	list


In [74]:
fruits.append('tomatoes')  # add an item to the end of the list
print(fruits)

['apples', 'bananas', 'pears', 'mangosteens', 'strawberries', 'tomatoes']


In [75]:
fruits.remove('tomatoes')  # remove the first item in the list that matches the argument
print(fruits)

['apples', 'bananas', 'pears', 'mangosteens', 'strawberries']


In [76]:
fruits.index('bananas')  # return the index of the first item matching the argument

1

In [77]:
fruits.append('bananas') # bring some extras in case we're hungry
fruits.count('bananas')  # return the number of times x appears in the list

2

In [78]:
eat = fruits.pop(0)  # return and remove item at position 0 from list (removes last item if no index is specified)
print(fruits)

['bananas', 'pears', 'mangosteens', 'strawberries', 'bananas']


In [79]:
print(eat)  # the item previously at position 0 ('apples') is now 'popped' into the variable eat

apples


Not exactly a method, but the `in` keyword is useful for checking if a list contains a particular item.

In [80]:
'bananas' in fruits

True

In [81]:
'tomatoes' in fruits # Not anymore!

False

## Dictionaries

The final data structure we'll talk about before taking a break are dictionaries. Dictionaries (or, more properly, `dicts`) work a lot like lists, except instead of looking up things in a dictionary by their _index_, we look them up by their _key_. It's easy to think of a `Python` dictionary as almost the same thing as a Webster's dictionary:
* __The Key:__ The "key" in a `Python` dictionary corresponds to the word you want to look up.
* __The Value:__ The "value" in a `Python` dictionary is whatever is stored at a key, which corresponds to a word's definition in a real dictionary.

(Dictionaries are known as *associative arrays* or *hash tables* in other languages.)

Let's try it. Much like a list, dictionaries are constructed using curly braces, with key value pairs separted by commas, like so: `{key1: value1, key2: value2, key3: value3}`. Let's try it:

In [82]:
gre_study_guide = {
    "bucolic": "adj. pastoral, rustic, countryfied",
    "tendentious": "adj. controversial, one-sided",
    "skulk": "v. to move in a stealthy or furtive manner"
}

Now, suppose the big test is tomorrow, but we can't remember the defintion of "skulk." No problem---we'll just look it up in much the same way that we'd look something up in a list:

In [83]:
gre_study_guide["skulk"]

'v. to move in a stealthy or furtive manner'

Of course, like lists, dictionaries can store more than strings. For instance, you might use a dictionary to store information for something like an address book:

In [84]:
address_book = {
    "simon": {
        "first_name": "Simon",
        "last_name": "Cowell",
        "phone": 447911123456
    },
    "paula": {
        "first_name": "Paula",
        "last_name": "Abdul",
        "phone": 5303228051
    },
    "randy": {
        "first_name": "Randy",
        "last_name": "Jackson",
        "phone": 2122002099
    }
}

Now we can find someone's contact information just by looking up their name.

In [85]:
address_book["randy"]

{'first_name': 'Randy', 'last_name': 'Jackson', 'phone': 2122002099}

What's more, we can even go directly to their phone number by accessing the `"phone"` key in the dictionary returned after looking up `"randy"`!

In [86]:
address_book["randy"]["phone"]

2122002099

Let's get familiar with how we can work with dictionaries.

In [87]:
me = {'name':'Hans', 'email':'jgaeb@stanford.edu'}
print(me)

{'name': 'Hans', 'email': 'jgaeb@stanford.edu'}


In [88]:
me['cell'] = '414-123-4567'
print(me)

{'name': 'Hans', 'email': 'jgaeb@stanford.edu', 'cell': '414-123-4567'}


Or delete existing `key:value` pairs with the `del` statement.

In [89]:
del(me['email'])
print(me)

{'name': 'Hans', 'cell': '414-123-4567'}


The `key` of a dictionary can't be a list (because lists are mutable), but the `value` sure can!

In [90]:
me['siblings'] = ['Carrie', 'Karl']
print(me)

{'name': 'Hans', 'cell': '414-123-4567', 'siblings': ['Carrie', 'Karl']}


Use the `keys()` method of dictionary objects to get a list of the keys used in the dictionary.

In [91]:
me.keys()

dict_keys(['name', 'cell', 'siblings'])

And use the `in` keyword (compatible with all lists) to see if the a certain key exists in the dictionary.

In [92]:
'name' in me.keys()

True

In [93]:
'email' in me.keys()

False

When the keys are simple strings, it is sometimes easier to specify pairs using the `dict` constructor.

In [94]:
me = dict(name='Hans', email='jgaeb@stanford.edu', siblings=['Carrie', 'Karl'])
print(me)

{'name': 'Hans', 'email': 'jgaeb@stanford.edu', 'siblings': ['Carrie', 'Karl']}


Of course, it's worth keeping in mind is that the `key` in a dictionary can be _anything_, as long as it's immutable. So `key`s can be strings, numbers, or... tuples!

## Tuples

The (very last!) data type we'll need to talk about in `Python` is tuples. Fortunately, tuples are much easier than lists or dicts. Think of tuples as lists that are _immutable_: once you've put some stuff in a tuple, you can't change it.

Tuples consist of a number of values separted by commas (not necessarily, but often, enclosed in parentheses).

In [95]:
description = 'male', 'dark hair'
print(description)

('male', 'dark hair')


In [96]:
description[0]  # tuples are also sequences, and can be indexed

'male'

In [97]:
description[1:]  # or sliced

('dark hair',)

In [98]:
description[0] = 'female'  # but NOT changed, because they are immutable

TypeError: 'tuple' object does not support item assignment

While being immutable may seem like a minor difference from lists, the implications are quite big, and tuples are generally used for very different purposes compared to lists. For example, tuples can be used as the `key` for dictionaries (think sparse matrices). 

In [99]:
super_sparse_matrix = {(0, 0):1, (1000, 1000):1}  # a 1000*1000 matrix with only two non-zero elements?
print(super_sparse_matrix)

{(0, 0): 1, (1000, 1000): 1}


In [100]:
word_matrix = {('apples', 'bananas'):1, ('apples', 'pears'):1}  # a matrix indexed by words
print(word_matrix)

{('apples', 'bananas'): 1, ('apples', 'pears'): 1}


There are many more data structures commonly used in `python`, but lists, dictionaries, and tuples pretty much cover the basics (not to mention that these three constitute enough to fully represent the [JSON](http://json.org/) format in `python`, something you might see some of this afternoon when you're working with APIs and scrapers.)

# Exercises

Try these over the break. Work with the people sitting around you.

## Exercise 1.
1. Declare a string variable named **`s`**, that has the value:
> "double quotes" and single 'quotes' are equally acceptable in python

1. Make `Python` count how many times the letter `t` appears in the string **`s`**.
1. Replace all quotation marks in the string **`s`** with an underbar ('\_').
 - to be precise, you're not *replacing* the quotations, but *reassigning* the variable **`s`** with a copy of the old **`s`** that has underbars replacing quotations; remember that `python` strings are **immutable** (i.e., they are NEVER changed, only reassigned).
 - this was not covered above, but you should use `python`'s `replace()` method for strings; now would be a good time to practice reading the docs (https://docs.python.org/3/library/stdtypes.html#str.replace).
1. Split the string **`s`** into a list named **`words`**.
1. Count the length of the string **`s`** and the list **`words`**.
1. Join the elements of the list **`words`** into a comma separated string.

## Exercise 2.
1. Create a very, very _long_ string that contains all of Charles Dickens's _A Tale of Two Cities_ using the following code. (Don't worry that this doesn't make much sense now---we'll get to what's going on here in the next lesson.) `with open("data/two_cities.txt", "r") as f: ttc = f.read()`
2. Turn `ttc` into a list by splitting on spaces (i.e., the string `" "`). How many words is _A Tale of Two Cities_?
3. How many times does the word "the" occur in _A Tale of Two Cities_? How about "king"? (__HINT:__ Use the `.count()` function. For instance, `[1,2,3,1].count(1) = 2`.)

## Exercise 3.
1. Turn `ttc` from the previous exercise into a dictionary using the following code. (Don't worry if this code doesn't make sense.) `ttc = {word:ttc.count(word) for word in ttc[:1000]}`
- This dictionary is made up of the following `key: value` pairs: the keys are the first 1000 words in _A Tale of Two Cities_, and the values are the number of times the word appears in the whole book.
- This might take a little while, so don't be alarmed if this doesn't happen right away.
2. Check how many times the word "lords" occurs.
3. Reduce the number of times a word you don't like occurs by modifying a `key: value` pair in `ttc`.
4. Add a word that you think _should_ occur but doesn't to `ttc`.